src/dataflow/src/sink/tail.rs

```diff
@@ -30,19 +35,29 @@ pub fn tail<G>(
     let mut tx = block_on(connector.tx.connect()).expect("tail transmitter failed");
     let mut packer = RowPacker::new();
     stream.sink(Pipeline, &format!("tail-{}", id), move |input| {
-        input.for_each(|_, rows| {
+        input.for_each(|_, batches| {
             let mut results: Vec<Row> = Vec::new();
-            for (row, time, diff) in rows.iter() {
-                let should_emit = if connector.strict {
-                    connector.frontier.less_than(time)
-                } else {
-                    connector.frontier.less_equal(time)
-                };
-                if should_emit {
-                    packer.push(Datum::Decimal(Significand::new(i128::from(*time))));
-                    packer.push(Datum::Int64(i64::cast_from(*diff)));
-                    packer.extend_by_row(row);
-                    results.push(packer.finish_and_reuse());
+            for batch in batches.iter() {
+                let mut cursor = batch.cursor();
+                while cursor.key_valid(&batch) {
+                    while cursor.val_valid(&batch) {
+                        let row = cursor.val(&batch);
+                        cursor.map_times(&batch, |time, diff| {
```

Comment:
This part confuses me. There's multiple (time, diff) tuples for each row? Does this ever have more than 1 entry? If so it does not guarantee the "timestamps are in order" property I'm trying to get here. Do I instead need to batch up all (row, time, diff) tuples in each `batch` and send those in order? Is each `batch`'s frontier guaranteed to be >= any previous frontier?

Reply:
Or I suppose I could do results.sort_by at the end, if the batches are guaranteed to be non-overlapping and increasing.

Reply:
Ah, sorry I was sculpting my comment for a few minutes and these showed up about essentially the same topic while I was writing.

There can be multiple `(time, diff)` tuples for each row, and they definitely can have more than one entry (a batch spans an interval of time, and the record may come, go, come again in that time).

At each worker, each batch is guaranteed to correspond to a disjoint interval of time that strictly increases. The `.lower` field of a batch's description should always match the `.upper` field of the previously received batch.